### PR TITLE
feature #3403 : Added --s2i flag for 'odo create' command

### LINKED ISSUE #3403: odo create: force s2i components and cleanup
/kind user-story

## User Story
As an odo v2 user, I want to be able to create old s2i builder image (LocalConfig) components so I can work as with odo v1.

## Acceptance Criteria 
- [ ] It should be possible to force `odo create` command to create LocalConfig component even if it has the component type with the same name as devfile component. For example, using a new flag (`odo create java --s2i`)
- [ ] If there is a devfile component with a given name but no s2i component `odo create --s2i` should fail. 
- [ ] flags that are valid only for s2i components should have a note in  help message ( `odo create --help`)
- [ ] if the user provides a flag that is valid only  s2i components without `--s2i` the command should error out
- [ ] remove `--cpu`, `--max-cpu`,`--max-memory`,`--memory`,`--max-memory`,`--min-memory` flags to simplify odo create command. They were rarely used and the same can be achieved using `odo config` command after the component is created.

### component creation
```
▶  odo preference set experimental true

▶ odo catalog list components
Odo OpenShift Components:
NAME              PROJECT       TAGS                        SUPPORTED
java              openshift     11,8,latest                 YES
nodejs            openshift     10-SCL,8,8-RHOAR,latest     YES

Odo Devfile Components:
NAME                 DESCRIPTION                            REGISTRY                   SUPPORTED
maven                Upstream Maven and OpenJDK 11          DefaultDevfileRegistry     YES
nodejs               Stack with NodeJS 10                   DefaultDevfileRegistry     YES

▶ odo create nodejs
# this currently creates nodejs devfile component, which is OK
# but there is no way to create nodejs OpenShift component

▶ odo create nodejs --s2i
# this new proposed flag would create nodejs OpenShift component 

```

### example of help output

```
Flags:
      --app string                                       Application, defaults to active application
  -b, --binary string                                    (valid only with --s2i) Create a binary file component component using given artifact. Works only with Java components. File needs to be in the context directory.
      --context string                                   Use given context directory as a source for component settings
      --devfile string                                   Path to the user specify devfile
      --downloadSource string[="devfile-project-name"]   Download sample project from devfile.
      --env strings                                      (valid only with --s2i) Environmental variables for the component. For example --env VariableName=Value
  -g, --git string                                       (valid only with --s2i) Create a git component using this repository.
  -h, --help                                             Help for create
      --now                                              Push changes to the cluster immediately
  -p, --port strings                                     (valid only with --s2i) Ports to be used when the component is created (ex. 8080,8100/tcp,9100/udp)
      --project string                                   Project, defaults to active project
  -r, --ref string                                       (valid only with --s2i) Use a specific ref e.g. commit, branch or tag of the git repository
      --registry string                                  Create devfile component from specific registry

```

/kind user-story
/priority medium

### PR DESCRIPTION
**Pull request by dev-gaur**
**What type of PR is this?**
/kind feature

**What does does this PR do / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #3403 

**How to test changes / Special notes to the reviewer**:
Use the --s2i flag when using `odo create` to force S2i component over a namesake devfile component type.

### CODE REVIEW

File: pkg/odo/cli/component/create.go [hunk 1]
```diff
@@ -353,8 +389,20 @@ func (co *CreateOptions) Complete(name string, cmd *cobra.Command, args []string
...
+			catalogList, err = catalog.ListComponents(client)
+			if err != nil {
+				return err
+			}
+		}
+
 		// Validate user specify devfile path
 		if co.devfileMetadata.devfilePath.value != "" {
+			if co.forceS2i {
+				return errors.New("You can't set --s2i flag as true if you want to use the devfile that is specified via --devfile")
```
Comment (adisky) [inline pkg/odo/cli/component/create.go:404]: can we check for all conflicting flags like `--registry, --starter, --token etc` at one place only.  I think this also could be done in the `checkConflictingFlags()` function above.
```
if co.forceS2i {
     if co.devfileMetadata.devfilePath.value  != ""  || co.devfileMetadata.devfileRegistry.Name != ""{ 
     return errors.New("You can't set --s2i flag as true if you want to use the devfile that is specified via --devfile")
}
```

File: pkg/odo/cli/component/create.go [hunk 2]
```diff
@@ -422,6 +474,15 @@ func (co *CreateOptions) Complete(name string, cmd *cobra.Command, args []string
 				// Component type: We provide devfile component list to let user choose
 				componentType = ui.SelectDevfileComponentType(catalogDevfileList.Items)
 
+				if co.forceS2i {
```
Comment (adisky) [inline pkg/odo/cli/component/create.go:477]: if i am doing `odo create --s2i` in interactive mode, it is still giving me list of devfile components.

File: pkg/odo/cli/component/create.go [hunk 3]
```diff
@@ -794,7 +859,7 @@ func (co *CreateOptions) Complete(name string, cmd *cobra.Command, args []string
 func (co *CreateOptions) Validate() (err error) {
 
 	if experimental.IsExperimentalModeEnabled() {
-		if co.devfileMetadata.devfileSupport {
+		if !useS2I && co.devfileMetadata.devfileSupport {
```
Comment (adisky) [inline pkg/odo/cli/component/create.go:862]: why we need `useS2I`? when we have` co.forceS2i`?

**Comment by adisky** [comment]
@dev-gaur I tried it working well, in two cases i have faced problem
- Interactive mode doing `odo create --s2i` listing devfile components
- (edge case) If a devfile.yaml is present in current directory and i still wants to create `odo create nodejs --s2i` then it is taking devfile giving error ` ✗  Cannot select this component with --s2i flag`.

**Comment by adisky** [comment]
we use fixes to link the corresponding github issue, no need to mention it in PR heading `
feature #3403 : Added --s2i flag for 'odo create' command, it could be simply  ` Added --s2i flag for 'odo create' command`
`

File: pkg/odo/cli/component/create.go [hunk 4]
```diff
@@ -1149,6 +1214,7 @@ func NewCmdCreate(name, fullName string) *cobra.Command {
 		componentCreateCmd.Flags().StringVar(&co.devfileMetadata.devfileRegistry.Name, "registry", "", "Create devfile component from specific registry")
 		componentCreateCmd.Flags().StringVar(&co.devfileMetadata.devfilePath.value, "devfile", "", "Path to the user specify devfile")
 		componentCreateCmd.Flags().StringVar(&co.devfileMetadata.token, "token", "", "Token to be used when downloading devfile from the devfile path that is specified via --devfile")
+		componentCreateCmd.Flags().BoolVar(&co.forceS2i, "s2i", false, "Path to the user specify devfile")
```
[the remark this answers is not in the document]
Comment (dev-gaur) [reply pkg/odo/cli/component/create.go:1217]: aah crap. I totally forgot :sweat_smile:

File: docs/dev/experimental-mode.adoc [hunk 5]
```diff
@@ -123,3 +123,29 @@ $ odo delete myspring
 The devfile registry can be accessed link:https://github.com/elsony/devfile-registry[here]. 
 
 For more information on how to develop and write a devfile, please read the link:https://docs.google.com/document/d/1piBG2Zu2PqPZSl0WySj25UouK3X5MKcFKqPrfC9ZAsc[Odo stack creation] document.
+
+###### Forcing s2i type component creation over a namesake devfile type component:
```
[the remark this answers is not in the document]
Comment (dev-gaur) [reply docs/dev/experimental-mode.adoc:127]: @adisky from what I perceived, `--s2i` flag is about conflict resolution. It's not about disabling devfile totally. The flag simply resolves the conflict when a s2i type component and devfile type component have the same name. for eg., nodejs component.

If a user picks a devfile type component without a conflicting name, then a devfile component shall be created despite --s2i flag. This is the same reason why I didn't disable interactive mode for devfile component type selection. 

@kadel @girishramnani  Can you confirm this? Please correct me if I'm wrong. I'll make the changes then.

File: pkg/odo/cli/component/create.go (hunk 1 above)
Comment (dev-gaur) [reply pkg/odo/cli/component/create.go:404]: The `checkConflictingFlags( )`  is actually there for the clause 
` if the user provides a flag that is valid only s2i components without --s2i the command should error out`

from https://github.com/openshift/odo/issues/3403#issue-643752720 .

Maybe I should rename it to checkConfictingS2IFlags

File: pkg/odo/cli/component/create.go (hunk 2 above)
Comment (dev-gaur) [reply pkg/odo/cli/component/create.go:477]: https://github.com/openshift/odo/pull/3545#discussion_r459221704

**Comment by dev-gaur** [comment]
> @dev-gaur I tried it working well, in two cases i have faced problem
> 
> * Interactive mode doing `odo create --s2i` listing devfile components

Explained in above comments.

> * (edge case) If a devfile.yaml is present in current directory and i still wants to create `odo create nodejs --s2i` then it is taking devfile giving error ` ✗  Cannot select this component with --s2i flag`.

Was that a nodejs component? If yes, then the behaviour is expected. Else, I'll make the changes.

> we use fixes to link the corresponding github issue, no need to mention it in PR heading ` feature #3403 : Added --s2i flag for 'odo create' command`, it could be simply `Add --s2i flag for 'odo create' command`

Noted. Its an old habit from the jkube team :sweat_smile:

File: pkg/odo/cli/component/create.go (hunk 1 above)
Comment (adisky) [reply pkg/odo/cli/component/create.go:407]: what i meant here you could check devfile and s2i conflicting flags at one place only, may be define another fuction `checkS2iDevfileConflictFlags`, i see you are checking each one at separate places.

File: docs/dev/experimental-mode.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (adisky) [reply docs/dev/experimental-mode.adoc:127]: @dev-gaur your understanding is correct here, but please check the format here https://github.com/openshift/odo/blob/cb24286544cc48dec5ccaa4be8fba25766047f1e/docs/dev/experimental-mode.adoc#forcing-s2i-type-component-creation-over-a-namesake-devfile-type-component
and this heading `Forcing s2i type component creation over a namesake devfile type component` is not easily understandable

File: pkg/odo/cli/component/create.go (hunk 2 above)
Comment (adisky) [reply pkg/odo/cli/component/create.go:480]: it is giving only devfile component list, not s2i component. IMO this flag should be disabled for interactive mode.

**Comment by adisky** [comment]
@dev-gaur I think we can disable `--s2i` flag for interactive mode, @kadel @girishramnani WDYT?

File: docs/dev/experimental-mode.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (dev-gaur) [reply docs/dev/experimental-mode.adoc:127]: will correct this

File: pkg/odo/cli/component/create.go (hunk 1 above)
Comment (dev-gaur) [reply pkg/odo/cli/component/create.go:407]: I thought its easier to follow the code flow this way.  https://en.wikipedia.org/wiki/KISS_principle

**Comment by girishramnani** [comment]
will take a look at this in the evening

**Comment by girishramnani** [comment]
> @dev-gaur I think we can disable `--s2i` flag for interactive mode, @kadel @girishramnani WDYT?

I mean cant we ask the user if you want to use `legacy` component type or something?

**Comment by cdrage** [comment]
> @dev-gaur I think we can disable `--s2i` flag for interactive mode, @kadel @girishramnani WDYT?

I don't see how this would be hard to implement. For interactive mode, it'd just check to see if `--s2i` has been passed in? We could do it so that interactive mode is either Devfile with `odo create` or s2i with `odo create --s2i`

**Comment by dev-gaur** [comment]
> > @dev-gaur I think we can disable `--s2i` flag for interactive mode, @kadel @girishramnani WDYT?
> 
> I don't see how this would be hard to implement. For interactive mode, it'd just check to see if `--s2i` has been passed in? We could do it so that interactive mode is either Devfile with `odo create` or s2i with `odo create --s2i`

ofcourse its not hard to implement. I just wanted to discuss if this should be done. And if yes, should --s2i be disabled in the favour of interactive mode or should interactive mode be disabled in favour of --s2i flag.

And now I got my answer from the above comments, the former option.

**Comment by adisky** [comment]
@cdrage sorry i have not explained myself clearly. It is not about skipping interactive mode for s2i, it is about interactive mode support with the flag `odo create --s2i`, as it looks confusing. There could be two ways of doing that:

1. `odo create --s2i` would be the interactive mode for s2i,  
cons: with interactive mode we don't specify a single flag, so this is against odo consistency.

2. `odo create` in interactive mode should we ask s2i or devfile? 
cons: we do not want to give first hand support for s2i components.
In this case we could print deprecation warning if user selects s2i.

IMO approach 2 is better,

**Comment by dev-gaur** [comment]
> odo create --s2i would be the interactive mode for s2i,
cons: with interactive mode we don't specify a single flag, so this is against odo consistency.

Can't we make an exception for experimental mode here? The mode is called "experimental" for a reason :wink:  

This approach feels neat to me, IMO

**Comment by girishramnani** [comment]
can we just ignore the `--s2i` flag on interactive mode and then have a follow up?

**Comment by dev-gaur** [comment]
> can we just ignore the `--s2i` flag on interactive mode and then have a follow up?

I'm not sure what do you mean by 'follow up' here. 

If you mean we let the user select the devfile component type in the interactive mode and if --s2i flag is enabled, then check for conflicting names and act in favour of s2i type , then this is the current approach I adopted.

**Comment by kadel** [comment]
> 1. `odo create --s2i` would be the interactive mode for s2i,
>    cons: with interactive mode we don't specify a single flag, so this is against odo consistency.

+1

**Comment by adisky** [comment]
> @cdrage sorry i have not explained myself clearly. It is not about skipping interactive mode for s2i, it is about interactive mode support with the flag `odo create --s2i`, as it looks confusing. There could be two ways of doing that:
> 

agreed on this +1

**Comment by cdrage** [comment]
> > 1. `odo create --s2i` would be the interactive mode for s2i,
> >    cons: with interactive mode we don't specify a single flag, so this is against odo consistency.
> 
> +1

triple agree on this. 

Having it inside of `odo create` to pick between Devfile / S2I seems like it'd only be in odo temporarily since we're encouraging the use of Devfile.

File: pkg/odo/cli/component/create.go [hunk 6]
```diff
@@ -359,8 +390,20 @@ func (co *CreateOptions) Complete(name string, cmd *cobra.Command, args []string
...
+			catalogList, err = catalog.ListComponents(client)
+			if err != nil {
+				return err
+			}
+		}
+
 		// Validate user specify devfile path
 		if co.devfileMetadata.devfilePath.value != "" {
+			if co.forceS2i {
+				return errors.New("You can't set --s2i flag as true if you want to use the devfile that is specified via --devfile")
```
Comment (cdrage) [inline pkg/odo/cli/component/create.go:405]: errors cannot be capitalized. Please remove the capitalization at the beginning.

File: pkg/odo/cli/component/create.go [hunk 7]
```diff
@@ -374,6 +417,10 @@ func (co *CreateOptions) Complete(name string, cmd *cobra.Command, args []string
 
 		// Validate user specify registry
 		if co.devfileMetadata.devfileRegistry.Name != "" {
+			if co.forceS2i {
+				return errors.New("You can't specify registry via --registry if you want to force the S2I build with --s2i")
```
Comment (cdrage) [inline pkg/odo/cli/component/create.go:421]: remove capitalization

File: pkg/odo/cli/component/create.go [hunk 8]
```diff
@@ -448,6 +495,10 @@ func (co *CreateOptions) Complete(name string, cmd *cobra.Command, args []string
 			if util.CheckPathExists(DevfilePath) || co.devfileMetadata.devfilePath.value != "" {
 				// Use existing devfile directly
 
+				if co.forceS2i {
+					return errors.Errorf("Existing devfile component detected. Please remove the devfile component before creating an s2i component")
```
Comment (cdrage) [inline pkg/odo/cli/component/create.go:499]: remove capitalization

File: pkg/odo/cli/component/create.go [hunk 9]
```diff
@@ -563,6 +614,19 @@ func (co *CreateOptions) Complete(name string, cmd *cobra.Command, args []string
...
+			if co.forceS2i && hasComponent {
+				s2iOverride := false
+				for _, item := range catalogList.Items {
+					if item.Name == co.devfileMetadata.componentType {
+						s2iOverride = true
+						break
+					}
+				}
+				if !s2iOverride {
+					return errors.New("Cannot select this component with --s2i flag")
```
Comment (cdrage) [inline pkg/odo/cli/component/create.go:626]: remove capitalization

File: pkg/odo/cli/component/create.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (cdrage) [reply pkg/odo/cli/component/create.go:1218]: `user specified devfile` sounds more grammatically correct :+1:

[the remark this answers is not in the document]
Comment (cdrage) [reply pkg/odo/cli/component/create.go:1218]: This needs to be updated so say something like: "Enable S2i support" or something similar.

**Comment by adisky** [comment]
@devang working well now, good to go from my side with address @cdrage comments addressed, and update the flag description.
Two things i have observed that could be taken care of in a followup PR, I am able to specify `token` and `starter` flags that are devfile specific, so you can print error message from them also as you did with `registry` flag. 
```odo create nodejs --s2i --token kdlkad```
```odo create nodejs --s2i --starter myproject```

**Comment by dev-gaur** [comment]
changes made.

**Comment by kadel** [comment]
> New changes are detected. LGTM label has been removed.

adding lgtm back after rebase
/lgtm

**Comment by cdrage** [comment]
Changes lgtm

/lgtm

**Comment by adisky** [comment]
some real failures
```
[Fail] odo devfile create command tests When executing odo create with component type argument and --s2i flag [It] should fail the create command as --now flag, which is specific to s2i component creation, is used without --s2i flag 
/home/travis/gopath/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_create_test.go:214
[Fail] odo devfile create command tests When executing odo create with devfile component type argument and --registry flag [It] should fail to create the devfile component if specified registry is invalid 
/home/travis/gopath/src/github.com/openshift/odo/tests/helper/helper_generic.go:67
[Fail] odo devfile create command tests When executing odo create with component type argument and --s2i flag [It] should fail the create command as --git flag, which is specific to s2i component creation, is used without --s2i flag 
/home/travis/gopath/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_create_test.go:201
[Fail] odo devfile create command tests When executing odo create with component type argument and --s2i flag [It] should fail the create command as --binary flag, which is specific to s2i component creation, is used without --s2i flag 
/home/travis/gopath/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_create_test.go:208
```
holding it until tests are not resolved, @dev-gaur remove `/hold` once you fix the tests. 

/hold

**Comment by cdrage** [comment]
Looks like a flake.

So going to 

/retest

**Comment by dev-gaur** [comment]
> some real failures
> 
> ```
> [Fail] odo devfile create command tests When executing odo create with component type argument and --s2i flag [It] should fail the create command as --now flag, which is specific to s2i component creation, is used without --s2i flag 
> /home/travis/gopath/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_create_test.go:214
> [Fail] odo devfile create command tests When executing odo create with devfile component type argument and --registry flag [It] should fail to create the devfile component if specified registry is invalid 
> /home/travis/gopath/src/github.com/openshift/odo/tests/helper/helper_generic.go:67
> [Fail] odo devfile create command tests When executing odo create with component type argument and --s2i flag [It] should fail the create command as --git flag, which is specific to s2i component creation, is used without --s2i flag 
> /home/travis/gopath/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_create_test.go:201
> [Fail] odo devfile create command tests When executing odo create with component type argument and --s2i flag [It] should fail the create command as --binary flag, which is specific to s2i component creation, is used without --s2i flag 
> /home/travis/gopath/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_create_test.go:208
> ```
> 
> holding it until tests are not resolved, @dev-gaur remove `/hold` once you fix the tests.
> 
> /hold

I ran the "odo devfile create command tests" suite and it is passing locally :confused:

**Comment by adisky** [comment]
@dev-gaur There is a extra string coming in CI, not matching with your expected output
```
• Failure [3.130 seconds]
odo devfile create command tests
/go/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_create_test.go:16
  When executing odo create with component type argument and --s2i flag
  /go/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_create_test.go:133
    should fail the create command as --git flag, which is specific to s2i component creation, is used without --s2i flag [It]
    /go/src/github.com/openshift/odo/tests/integration/devfile/cmd_devfile_create_test.go:199
    Expected
      
        <string>: I0803 10:42:03.604022    9039 preference.go:196] The path for preference file is /tmp/384490129/config.yaml
         ✗  flag --git, requires --s2i flag to be set true, when in experimental mode.
        
    to contain substring
        <string>: flag --git, requires --s2i flag to be set, when in experimental mode.
```

**Comment by dev-gaur** [comment]
image

Not sure If my changes were leading to this failure.

Nevertheless I have modified this test case in https://github.com/openshift/odo/pull/3545/commits/4a9691e20ccb9fe358ef1596074eb7aa1f33b1d1

**Comment by cdrage** [comment]
Tests pass on OpenShift CI

/lgtm

:+1:

**Comment by girishramnani** [comment]
@dev-gaur why does this PR show the old clusters?

**Comment by girishramnani** [comment]
@prietyc123 please report this to the ci platform team

**Pull request merged**